Switch to rattler-build

h-vetinari opened the pull request:
This PR is not immediately meant for merging, but for doing a logical separation of different changes on the way to exploratory emscripten support. Builds on #82 

Still needs a bunch of fixes in our infra.

h-vetinari committed:
switch to CMake + ninja on unix

h-vetinari committed:
switch to rattler-build

c.f. https://prefix.dev/blog/rattler_build_on_conda_forge

h-vetinari committed:
alphabetize conda-forge.yml

h-vetinari committed:
MNT: Re-rendered with conda-build 24.7.1, conda-smithy 3.39.2.dev18+gf7eb69ba, and conda-forge-pinning 2024.09.09.04.46.35

h-vetinari committed:
pass `--experimental` also to build invocation

h-vetinari committed:
reapply "pass `--experimental` also to build invocation"

h-vetinari committed:
reapply "pass `--experimental` also to build invocation"

h-vetinari committed:
specify script for cache phase

h-vetinari committed:
rename build-scripts for global stage

Suggested-By: Wolf Vollprecht <w.vollprecht@gmail.com>

h-vetinari committed:
folderize globs

wolfv commented on recipe/recipe.yaml:
```diff
@@ -93,8 +93,8 @@ outputs:
+            - lib/pkgconfig/
             - lib/libz.a
           else:
             - Library/include
```
you probably need to folderize these, too :)

h-vetinari committed:
ensure LIBRARY_{BIN,LIB} exist before copying stuff there

h-vetinari commented:
There's a subtle behaviour change in how `%LIBRARY_LIB%` existed during the global build stage before, but doesn't anymore. No big deal, but triggered a failed copy here, c.f. https://github.com/conda-forge/zlib-feedstock/pull/83/commits/70e47af7580eec0cffc6e42ec026d1af6830bfd6

wolfv commented:
I've re-triggered some builds to see if `rattler-build` 0.22.0 changes anything :)

h-vetinari commented:
Interesting - on windows, `cmake` doesn't seem to be found; it's definitely there though, and also works before the translation to RB. 
```
 │ │ Resolving build environment:
 │ │   Platform: win-64 [__win=0=0, __archspec=1=x86_64_v4]
 │ │   Channels: 
 │ │    - file:///D:/bld/
 │ │    - conda-forge
 │ │   Specs:
 │ │    - cmake
 │ │    - ninja
 │ │    - vs2019_win-64
 │ │    - vs_win-64
 │ │ [...]
 │ │ 'cmake' is not recognized as an internal or external command,
 │ │ operable program or batch file.
```

wolfv commented:
We can see that the `osx` version is detected as 10.16 now, too: 

```
 │ │ Resolving host environment:
 │ │   Platform: osx-64 [__unix=0=0, __osx=10.16=0, __archspec=1=skylake]
 │ │   Channels: 
 │ │    - file:///Users/runner/miniforge3/conda-bld/
 │ │    - conda-forge
 │ │   Specs:
 │ │    - __osx >=11.0
```

Which breaks the tests, since they require 11.0.

wolfv commented:
Strange because `jolt-physics` seems fine (also CMake) https://github.com/conda-forge/jolt-physics-feedstock/pull/3

h-vetinari commented:
> Strange because `jolt-physics` seems fine (also CMake) [conda-forge/jolt-physics-feedstock#3](https://github.com/conda-forge/jolt-physics-feedstock/pull/3)

Perhaps another little 🐞 hiding in `cache:`? 🤔

h-vetinari commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you but ran into some issues. Please check the output logs of the GitHub actions workflow below for more details. You can also ping conda-forge/core for further assistance or you can try [rerendering locally](https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-with-conda-smithy-locally.


The following suggestions might help debug any issues:
* Is the `recipe/{meta.yaml,recipe.yaml}` file valid?
* If there is a `recipe/conda-build-config.yaml` file in the feedstock make sure that it is compatible with the current [global pinnnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).
* Is the fork used for this PR on an organization or user GitHub account? Automated rerendering via the webservices admin bot only works for user GitHub accounts.


<sub>This message was generated by GitHub actions workflow run [https://github.com/conda-forge-webservices/actions/runs/11139428595](https://github.com/conda-forge-webservices/actions/runs/11139428595).</sub>

h-vetinari committed:
MNT: Re-rendered with conda-build 24.9.0, conda-smithy 3.41.1, and conda-forge-pinning 2024.10.01.17.37.14

h-vetinari commented:
FYI @wolfv @beckermr 
The admin bot didn't manage to rerender here, but manually it worked (modulo changes necessary for https://github.com/conda-forge/conda-smithy/issues/2057)

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/recipe.yaml```) and found some lint.

Here's what I've got...


For **recipe/recipe.yaml**:

 * ❌ The top level meta key cache is unexpected


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/12372865903](https://github.com/conda-forge/conda-forge-webservices/actions/runs/12372865903). Examine the logs at this URL for more detail.</sub>

h-vetinari committed:
lint: extraneous line at the end

h-vetinari commented:
> but manually it worked

Nevermind, that's almost certainly because I monkey-patched my local install (xref https://github.com/prefix-dev/rattler-build-conda-compat/issues/56).

h-vetinari commented:
@wolfv, `cmake` isn't found in _any_ of the builds -- at least now the error is consistent across platforms I guess 😅

Looks like a `cache:` problem to me. Do you want me to open an issue in rattler-build?

wolfv commented:
An issue with a small reproducer would be great :)

h-vetinari commented:
> An issue with a small reproducer would be great :)

The state of this PR is a reproducer, and already pretty minimal. 🙃 

Obviously you can leave out all the outputs, as it fails on the first invocation of `cmake` during the `cache:` stage, but that's purely deleting stuff, so not really strenuous I'd say.

h-vetinari commented:
_Now_ it should be pretty minimal. ;-)

beckermr commented:
@conda-forge-admin relint

h-vetinari commented:
> _Now_ it should be pretty minimal. ;-)

And now it's _slightly_ less minimal, but actually reproduces the problem 😅

h-vetinari committed:
DEBUG: disable output validation

Another issue with missing `--experimental` in ci-setup:
https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/367229158acb6873003058534ccfdc03976f8ec4/recipe/conda_forge_ci_setup/utils.py#L76-L83

h-vetinari commented:
@wolfv, thanks for the new rattler-build release! Looking better now (had to disable interactions with ci-setup, which cannot handle `--experimental` yet either). There was a segfault on linux-aarch64 that went away on restart, but otherwise unix looks fine.

On `win-64` the tests fail with some unclear cause
```
 │ │ │ Installing test environment
 │ │ │  Successfully updated the test environment
 │ │ │ Testing commands:
 │ │ │ (base) C:\Users\VSSADM~1\AppData\Local\Temp\.tmpdO7zDL>IF "" == "" (call C:\Users\VSSADM~1\AppData\Local\Temp\.tmpdO7zDL\build_env.bat ) 
 │ │ │
 │ │ ╰─────────────────── (took 0 seconds)
 │ │
 │ ╰─────────────────── (took 0 seconds)
 │
 ╰─────────────────── (took 71 seconds)
 × error Error building package: failed to run test: Script failed with status Some(1).
 × error Work directory: "C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\.tmpdO7zDL"
 × error To debug the build, run it manually in the work directory (execute the `./conda_build.sh` or `conda_build.bat` script)
Error:   × failed to run test: Script failed with status Some(1).
```

and on `win-arm64` it crashes as before, but now we see why
```
thread 'main' panicked at src\cache.rs:281:9:
assertion `left == right` failed: Prefixes must have the same length: "D:\\bld\\bld\\rattler-build_libzlib_1727940322\\h_env" != "D:\\bld\\bld\\rattler-build_libzlib-wapi_1727940322\\h_env"
  left: 49
 right: 54
```
Difference is related to the outputs somehow, i.e. `rattler-build_libzlib{,-wapi}_1727940322` in the environment path.

wolfv commented:
Yes, awesome. Thanks for retrying! On win-64 it fails during test-execution if I see correctly. The win-arm64 issue is curious. Will have to investigate (but looks fixable!).

h-vetinari commented:
> On win-64 it fails during test-execution if I see correctly.

Is this how test failures generally look, or are we running into specific case here? If it's due to a failing test (and not some problem in the supporting infra), I'd want to have _much_ more information about which line or assertion failed, rather than `Some(1)`.

> The win-arm64 issue is curious. Will have to investigate (but looks fixable!).

Cool, thanks!

wolfv commented:
@h-vetinari if you have an idea about a different way of invoking `cmd.exe` that would be welcome :) Maybe there are some parameters that will print the line that is executed etc. (like bash).

h-vetinari commented:
> @h-vetinari if you have an idea about a different way of invoking `cmd.exe` that would be welcome :) Maybe there are some parameters that will print the line that is executed etc. (like bash).

Just execute the individual instructions from the test section line-by-line?

wolfv commented:
I don't think that's the best solution as that's pretty different to other script executions. If it's in a `list` then it would be possible (otherwise we'd have to parse the script and split it on logical boundaries). I'll try to see if there is a way to enable a better way to trace script execution.

Btw. for package content tests, you can use the new `package_content` test that produce _much_ better output and run insanely fast (because they only check the `paths.json` file) :)

Example: https://github.com/conda-forge/jolt-physics-feedstock/blob/be2718fdc372f0bd5cb33950751eb7eaf092a591/recipe/recipe.yaml#L54-L59

h-vetinari commented:
> I don't think that's the best solution as that's pretty different to other script executions.

I think this is what conda-build does (but don't know); I'm not talking about the content of some `.cmd` or `.bat` script by the way, just the stuff that's under:
```
test:
  commands:
    - ...
```
should be executed one-by-one (and that principle can even apply to all platforms), because you really want to know at which line things failed, and not just "in the test section".

> Btw. for package content tests, you can use the new `package_content` test that produce _much_ better output

Sure, that'll cover most cases I hope. But wherever we run things that cannot be translated to other tests under `commands:`, those lines should be executed separately. They're already a list when they come out of the yaml, so you'd just have to iterate through.

h-vetinari commented:
> you can use the new `package_content` test

It also doesn't work here because we have a lib that doesn't follow the standard naming pattern (`z.lib`, `zlib.lib`, `zlib.dll`, `zlibstatic.lib`), and more importantly, the DLL and the import library are in separate outputs, and also there's a DLL in a non-standard location (not `%LIBRARY_BIN%`).

wolfv commented:
You can still use globs to check etc. make them platform independent and so on. All these cases are easily covered.

wolfv commented:
https://prefix-dev.github.io/rattler-build/latest/reference/recipe_file/#check-for-package-contents

h-vetinari commented:
Thanks, I guess using `files:` directly will work for presence tests.

> You can still use globs to check etc. make them platform independent and so on. All these cases are easily covered.

However, we often want to specifically check the absence of something, mostly that there's no static library. I don't see how that's covered yet.

wolfv commented:
Yes, we already have the design for negative globs as well :) https://github.com/prefix-dev/rattler-build/issues/915

h-vetinari commented:
I'll try to switch to the `package_content` tests, but regardless of how much can be converted, having no output for
```
test:
  commands:
    - ...
```
is a complete dealbreaker from my POV. I don't even see the inconsistency you mentioned with other script executions (well, those that are defined in a yaml list at least) - in those cases it's equally important to know which line failed.

wolfv commented:
Definitely! If you have a simple repro I can debug or can come up with something myself

h-vetinari committed:
MNT: Re-rendered with conda-build 24.9.0, conda-smithy 3.43.2, and conda-forge-pinning 2024.11.01.21.49.52

h-vetinari committed:
get rid of quotes in tests.script for libzlib-wapi

h-vetinari committed:
ensure test.script blocks are indented

h-vetinari committed:
switch to package_contents tests where possible

h-vetinari committed:
separate test block for package_contents

h-vetinari committed:
move `libz${SHLIB_EXT}` to lib-tests

h-vetinari committed:
can't have an empty script section, apparently

h-vetinari committed:
MNT: Re-rendered with conda-build 24.11.2, conda-smithy 3.45.1.dev2+g24c7f0a0, and conda-forge-pinning 2024.12.17.11.10.14

h-vetinari committed:
move source to cache
